### PR TITLE
feat: add endpoints for upload, list and delete

### PR DESCRIPTION
**Pull request by psankhe28**
## Summary by Sourcery

This pull request introduces new endpoints for file upload, list, and delete functionalities using Flask and Minio. It also adds a Makefile for managing various tasks, sets up GitHub Actions workflows for CI/CD, and configures Sphinx for documentation. Additionally, it includes initial test setups and various configuration files for code quality and security.

- **New Features**:
    - Added endpoints for file upload, list, and delete functionalities using Flask and Minio.
    - Introduced a new Makefile for managing environment, code quality, testing, and documentation tasks.
- **Enhancements**:
    - Implemented a TUS server for handling file uploads and assembly.
    - Configured Sphinx for generating project documentation.
- **Build**:
    - Added a Makefile to streamline build and environment management tasks.
- **CI**:
    - Set up GitHub Actions workflows for code quality checks, testing, documentation, and vulnerability scanning.
    - Added a workflow for automatic project structure updates using Cruft.
- **Deployment**:
    - Configured a GitHub Actions workflow for publishing the package to PyPI upon release.
- **Documentation**:
    - Added Sphinx configuration and Makefile for generating project documentation.
    - Included a Code of Conduct document.
- **Tests**:
    - Added unit and integration test directories and initial test setup.
    - Configured GitHub Actions workflows for running unit and integration tests and uploading coverage reports.
- **Chores**:
    - Added a .pre-commit-config.yaml file for managing pre-commit hooks.
    - Included a .safety-policy.yml file for managing security policies.

### CODE REVIEW

**Comment by JaeAeich** [comment]
@psankhe28 dw about test and documentation CI, I'll fix that. for types, you need to install the types of the libraries you are using. Check the CI logs, or below.
```
Run poetry run mypy TusStorageHandler/
TusStorageHandler/tus_server.py:5: error: Cannot find implementation or library stub for module named "minio.error"  [import-not-found]
TusStorageHandler/minio_client.py:23: error: Cannot find implementation or library stub for module named "dotenv"  [import-not-found]
TusStorageHandler/minio_client.py:24: error: Cannot find implementation or library stub for module named "minio"  [import-not-found]
TusStorageHandler/app.py:7: error: Cannot find implementation or library stub for module named "flask"  [import-not-found]
TusStorageHandler/app.py:8: error: Library stubs not installed for "flask_cors"  [import-untyped]
TusStorageHandler/app.py:8: note: Hint: "python3 -m pip install types-Flask-Cors"
TusStorageHandler/app.py:8: note: (or run "mypy --install-types" to install all missing stub packages)
TusStorageHandler/app.py:9: error: Cannot find implementation or library stub for module named "minio.error"  [import-not-found]
TusStorageHandler/app.py:9: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
TusStorageHandler/app.py:[1](https://github.com/elixir-cloud-aai/tus-storagehandler/actions/runs/9723599468/job/26838898556?pr=1#step:4:1)2: error: Cannot find implementation or library stub for module named "werkzeug.utils"  [import-not-found]
Found 7 errors in 3 files (checked 3 source files)
```

File: .cruft.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
+{
+  "template": "D:\\open-source\\gsoc\\cookiecutter-python",
```
Comment (JaeAeich) [inline .cruft.json:2]: @uniqueg we need to merge and finalize the cookiecutter, when generated via gh, this shoul dbe gh link, that is not possible before merge.

File: .cruft.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{
+  "template": "D:\\open-source\\gsoc\\cookiecutter-python",
+  "commit": "7962344936dd1d331da1495ea5f258d18b7c883a",
+  "checkout": null,
+  "context": {
+    "cookiecutter": {
+      "author_name": "Elixir Cloud AAI",
+      "author_email": "cloud-service@elixir-europe.org",
+      "development_status": "Beta",
+      "short_description": "A short description of the project.",
```
Comment (JaeAeich) [inline .cruft.json:10]: :)

File: TusStorageHandler/minio_client.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      - MINIO_BUCKET_NAME
+
+    - Import `minio_client` to access:
+      - `minio_client`: Initialized Minio client instance
+      - `create_bucket()`: Function to create a new bucket if not exists.
+"""
+
+import os
+
+from dotenv import load_dotenv
```
Comment (JaeAeich) [inline TusStorageHandler/minio_client.py:23]: I personally think, using dotenv package, and letting python handle env variable is a good idea. I think environment variables should be handled by the "environment", expecting having a `.env` file with creds might be a bad idea. 

What I do is I use either my `bash/zsh{rc}`, export the variable in the shell env or just use a tool that help me manage my env variables, like [direnv](https://direnv.net/). Not asking you to set them up, or change your workflow just think `.env` file might not be ideal imo.

thoughts @uniqueg ??

File: TusStorageHandler/minio_client.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+from minio import Minio
+
+load_dotenv()
+
+MINIO_HOSTNAME = os.getenv("MINIO_HOSTNAME")
```
Comment (JaeAeich) [inline TusStorageHandler/minio_client.py:28]: I would get these in a seperate file like consts..py or env.py just so they are easier to document and then import it whereever needed.

File: TusStorageHandler/tus_server.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+  def get_file_path(self, upload_id):
+    """Returns the file path for the given upload_id within the 'uploads' directory.
+
+    Args:
+        upload_id (str): Unique identifier for the upload.
+
+    Returns:
+        str: File path within the 'uploads' directory corresponding to the upload_id.
+
```
Comment (JaeAeich) [inline TusStorageHandler/tus_server.py:35]: whats up with spaces after pydocstring, ruff won't pick these up ig as they don't violate with googles pydocstring guide (or I think so), but either be consistent and leave spaces in all or dont in any, I prefer latter :).

File: TusStorageHandler/tus_server.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    Args:
+        upload_id (str): Unique identifier for the upload location.
+        file_content (bytes): Content to be written to the file.
+
+    Raises:
+        IOError: If there is an error writing to the file.
+
+    """
+    file_path = self.get_file_path(upload_id)
+    if not os.path.exists(file_path):
```
Comment (JaeAeich) [inline TusStorageHandler/tus_server.py:51]: use `Path` from `pathlib`.
ie

```Python
from pathlib import Path

file_path = Path('path/to/your/file')

# Check if the path exists
if not file_path.exists():
    print("The path does not exist.")
else:
    print("The path exists.")
```

File: pyproject.toml [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+[tool.poetry.group.test.dependencies]
+pytest = ">=8.2.2,<9.0.0"
+pytest-cov = ">=5.0.0,<=6.0.0"
+
+[tool.poetry.group.types.dependencies]
+mypy = ">=1.10.0,<2.0.0"
+
+[tool.poetry.scripts]
+TusStorageHandler = "TusStorageHandler.main:main"
```
Comment (JaeAeich) [inline pyproject.toml:58]: there is no console script in `main.py`, main file doesn't even exist, remove this.

File: TusStorageHandler/app.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+MINIO_BUCKET = os.getenv("MINIO_BUCKET_NAME")
+
+app = Flask(__name__)
+CORS(app, resources={r"/*": {"origins": "http://localhost:3000"}})
+
+tus_server = TusServer(minio_client, MINIO_BUCKET)
+
+
+@app.route("/", methods=["GET"])
```
Comment (JaeAeich) [inline TusStorageHandler/app.py:22]: If you're writing an API why not use [FOCA](https://github.com/elixir-cloud-aai/foca/)??

File: TusStorageHandler/requirements.txt
Comment (JaeAeich) [inline TusStorageHandler/requirements.txt:1]: what is this??

Comment (psankhe28) [reply TusStorageHandler/requirements.txt:1]: The contents of the requirements.txt file are not showing up in the list of modified files, but they are visible when viewing the file in the branch. I'm not sure why this is happening.

File: TusStorageHandler/app.py (hunk 8 above)
Comment (psankhe28) [reply TusStorageHandler/app.py:22]: Yes, I plan to upgrade it using FOCA. First, I want to create a basic template to test it.

**Comment by JaeAeich** [comment]
use snake case in directory name.

**Review by uniqueg — CHANGES_REQUESTED** [review]
Thanks a lot, @psankhe28. Unfortunately, this PR is way too big for us to review effectively. Please break it down into several smaller PRs:
- 1 `chore` for `.gitignore`, `CODE_OF_CONDUCT.md`, `PULL_REQUEST_TEMPLATE.md` etc.
- 1 `'ci` for CI workflows
- 1 `docs` for docs stuff
- 1 `build` for build system (`pyproject.toml`)
- 3 `feat` for code (1 for each endpoint), each with unit tests and other corresponding changes, if applicable (e.g., doc/build updates); e.g., you don't wanna add dependencies (e.g., `tus`) to your previous `build` PR that you don't need at the time of adding the build system; only add them to the PR where they are needed

**Comment by uniqueg** [comment]
See https://github.com/elixir-cloud-aai/tus-storagehandler/pull/3#issuecomment-2250167889 and the comment above, @JaeAeich & @psankhe28

**Pull request closed without merging**